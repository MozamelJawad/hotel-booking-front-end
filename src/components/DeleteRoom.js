import React from 'react';
import { useSelector, useDispatch } from 'react-redux';
import { listRooms, deleteRoom } from '../redux/rooms/roomSlice';

const DeleteRoom = () => {
  const rooms = useSelector((state) => state.rooms.rooms);
  const categories = useSelector((state) => state.category.categories);
  const loading = useSelector((state) => state.rooms.loading);
  const error = useSelector((state) => state.rooms.error);
  const dispatch = useDispatch();

  const roomss = rooms.map((r) => ({
    ...r,
    category_name: categories.find((c) => c.id === r.category_id).name,
    price: categories.find((c) => c.id === r.category_id).price,
  }));

  const handleDelete = (id) => () => {
    dispatch(deleteRoom(id)).then(() => {
      dispatch(listRooms());
    });
  };

  if (loading) {
    return (
      <div className="div-center">
        <h3 className="text-center text-info text-wrap">loading ...</h3>
      </div>
    );
  }

  if (error) {
    return (
      <div className="div-center">
        <h3 className="text-center text-danger text-wrap">
          Fix the Error:
          {error }
        </h3>
      </div>
    );
  }

  return (
    <div className="container-fluid mt-5">
      <div className="table-responsive col-md-8 offset-md-2">
        <table className="table table-striped">
          <thead className="table-danger">
            <tr>
              <th scope="col">Room</th>
              <th scope="col">Price</th>
              <th scope="col">Category</th>
              <th scope="col">Action</th>
            </tr>
          </thead>
          <tbody>
            { roomss.map((room) => (
              <tr key={room.id}>
                <td>{room.name}</td>
                <td>{room.price}</td>
                <td>{room.category_name}</td>
                <td><button type="button" onClick={handleDelete(room.id)} className="btn btn-danger btn-sm"> Delete </button></td>
              </tr>
            ))}
          </tbody>
        </table>
      </div>
    </div>
  );
};

export default DeleteRoom;
