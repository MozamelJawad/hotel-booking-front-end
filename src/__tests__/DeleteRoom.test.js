import React from 'react';
// eslint-disable-next-line import/no-extraneous-dependencies
import renderer from 'react-test-renderer';
import { useSelector, useDispatch } from 'react-redux';
import DeleteRoom from '../components/DeleteRoom';

jest.mock('react-redux', () => ({
  ...jest.requireActual('react-redux'),
  useSelector: jest.fn(),
  useDispatch: jest.fn(),
}));

describe('Tests DeleteRoom component', () => {
  it('should render loading spinner when loading is true', () => {
    const loading = true;
    const error = null;
    const dispatchMock = jest.fn();

    useSelector.mockImplementation((selector) => selector({
      rooms: { rooms: [], loading, error },
      category: { categories: [] },
    }));

    useDispatch.mockReturnValue(dispatchMock);

    const DeleteRoomComponent = renderer.create(<DeleteRoom />);
    expect(DeleteRoomComponent.toJSON()).toMatchSnapshot();
  });

  it('should render error message when error is present', () => {
    const loading = false;
    const error = 'Something went wrong!';
    const dispatchMock = jest.fn();

    useSelector.mockImplementation((selector) => selector({
      rooms: { rooms: [], loading, error },
      category: { categories: [] },
    }));

    useDispatch.mockReturnValue(dispatchMock);

    const DeleteRoomComponent = renderer.create(<DeleteRoom />);
    expect(DeleteRoomComponent.toJSON()).toMatchSnapshot();
  });

  it('should render table with room data when rooms and categories are present', () => {
    const rooms = [
      { id: 1, name: 'Room 1', category_id: 1 },
      { id: 2, name: 'Room 2', category_id: 2 },
    ];
    const categories = [
      { id: 1, name: 'Category 1', price: 100 },
      { id: 2, name: 'Category 2', price: 200 },
    ];
    const loading = false;
    const error = null;
    const dispatchMock = jest.fn();

    useSelector.mockImplementation((selector) => selector({
      rooms: { rooms, loading, error },
      category: { categories },
    }));

    useDispatch.mockReturnValue(dispatchMock);

    const DeleteRoomComponent = renderer.create(<DeleteRoom />);
    expect(DeleteRoomComponent.toJSON()).toMatchSnapshot();
  });

  it('should render no rooms available message when rooms is empty', () => {
    const rooms = [];
    const categories = [
      { id: 1, name: 'Category 1', price: 100 },
      { id: 2, name: 'Category 2', price: 200 },
    ];
    const loading = false;
    const error = null;
    const dispatchMock = jest.fn();

    useSelector.mockImplementation((selector) => selector({
      rooms: { rooms, loading, error },
      category: { categories },
    }));

    useDispatch.mockReturnValue(dispatchMock);

    const DeleteRoomComponent = renderer.create(<DeleteRoom />);
    expect(DeleteRoomComponent.toJSON()).toMatchSnapshot();
  });
});
