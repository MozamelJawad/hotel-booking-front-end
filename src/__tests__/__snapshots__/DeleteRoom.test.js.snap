// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DeleteRoom component renders correctly 1`] = `
<DocumentFragment>
  <div
    class="container-fluid mt-5"
  >
    <div
      class="table-responsive col-md-8 offset-md-2"
    >
      <table
        class="table table-striped"
      >
        <thead
          class="table-danger"
        >
          <tr>
            <th
              scope="col"
            >
              Room
            </th>
            <th
              scope="col"
            >
              Price
            </th>
            <th
              scope="col"
            >
              Category
            </th>
            <th
              scope="col"
            >
              Action
            </th>
          </tr>
        </thead>
        <tbody>
          <tr>
            <td>
              Room 1
            </td>
            <td>
              100
            </td>
            <td>
              Category 1
            </td>
            <td>
              <button
                class="btn btn-danger btn-sm"
                type="button"
              >
                Delete
              </button>
            </td>
          </tr>
          <tr>
            <td>
              Room 2
            </td>
            <td>
              150
            </td>
            <td>
              Category 2
            </td>
            <td>
              <button
                class="btn btn-danger btn-sm"
                type="button"
              >
                Delete
              </button>
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Tests DeleteRoom component should render correctly 1`] = `
<div
  className="container-fluid mt-5"
>
  <div
    className="table-responsive col-md-8 offset-md-2"
  >
    <table
      className="table table-striped"
    >
      <thead
        className="table-danger"
      >
        <tr>
          <th
            scope="col"
          >
            Room
          </th>
          <th
            scope="col"
          >
            Price
          </th>
          <th
            scope="col"
          >
            Category
          </th>
          <th
            scope="col"
          >
            Action
          </th>
        </tr>
      </thead>
      <tbody>
        <tr>
          <td>
            Room 1
          </td>
          <td>
            100
          </td>
          <td>
            Category 1
          </td>
          <td>
            <button
              className="btn btn-danger btn-sm"
              onClick={[Function]}
              type="button"
            >
              Delete
            </button>
          </td>
        </tr>
        <tr>
          <td>
            Room 2
          </td>
          <td>
            200
          </td>
          <td>
            Category 2
          </td>
          <td>
            <button
              className="btn btn-danger btn-sm"
              onClick={[Function]}
              type="button"
            >
              Delete
            </button>
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;

exports[`Tests DeleteRoom component should render error message when error is present 1`] = `
<div
  className="div-center"
>
  <h3
    className="text-center text-danger text-wrap"
  >
    Fix the Error:
    Something went wrong!
  </h3>
</div>
`;

exports[`Tests DeleteRoom component should render loading spinner when loading is true 1`] = `
<div
  className="d-flex justify-content-center"
>
  <div
    className="text-success spinner-border mt-5"
    role="status"
  >
    <span
      className="visually-hidden"
    >
      Loading...
    </span>
  </div>
</div>
`;

exports[`Tests DeleteRoom component should render no rooms available message when rooms is empty 1`] = `
<div
  className="container-fluid mt-5"
>
  <div
    className="table-responsive col-md-8 offset-md-2"
  >
    <table
      className="table table-striped"
    >
      <thead
        className="table-danger"
      >
        <tr>
          <th
            scope="col"
          >
            Room
          </th>
          <th
            scope="col"
          >
            Price
          </th>
          <th
            scope="col"
          >
            Category
          </th>
          <th
            scope="col"
          >
            Action
          </th>
        </tr>
      </thead>
      <tbody />
    </table>
  </div>
</div>
`;

exports[`Tests DeleteRoom component should render table with room data when rooms and categories are present 1`] = `
<div
  className="container-fluid mt-5"
>
  <div
    className="table-responsive col-md-8 offset-md-2"
  >
    <table
      className="table table-striped"
    >
      <thead
        className="table-danger"
      >
        <tr>
          <th
            scope="col"
          >
            Room
          </th>
          <th
            scope="col"
          >
            Price
          </th>
          <th
            scope="col"
          >
            Category
          </th>
          <th
            scope="col"
          >
            Action
          </th>
        </tr>
      </thead>
      <tbody>
        <tr>
          <td>
            Room 1
          </td>
          <td>
            100
          </td>
          <td>
            Category 1
          </td>
          <td>
            <button
              className="btn btn-danger btn-sm"
              onClick={[Function]}
              type="button"
            >
              Delete
            </button>
          </td>
        </tr>
        <tr>
          <td>
            Room 2
          </td>
          <td>
            200
          </td>
          <td>
            Category 2
          </td>
          <td>
            <button
              className="btn btn-danger btn-sm"
              onClick={[Function]}
              type="button"
            >
              Delete
            </button>
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;
